opnode/l2/driver.go

```diff
@@ -0,0 +1,208 @@
+	l2Finalized eth.BlockID
+
+	// The L1 block we are syncing towards, may be ahead of l1Head
+	l1Target eth.BlockID
+
+	// Genesis starting point
+	Genesis Genesis
+}
+
+// L1Head returns the block-id (hash and number) of the last L1 block that was derived into the L2 block
+func (e *EngineDriver) L1Head() eth.BlockID {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l1Head
+}
+
+// L2Head returns the block-id (hash and number) of the L2 chain head
+func (e *EngineDriver) L2Head() eth.BlockID {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l2Head
+}
+
+func (e *EngineDriver) Head() (l1Head eth.BlockID, l2Head eth.BlockID) {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l1Head, e.l2Head
+}
+
+func (e *EngineDriver) UpdateHead(l1Head eth.BlockID, l2Head eth.BlockID) {
+	e.headLock.Lock()
+	defer e.headLock.Unlock()
+	e.l1Head = l1Head
+	e.l2Head = l2Head
+}
+
+func (e *EngineDriver) RequestHeadUpdate(ctx context.Context) error {
+	e.headLock.Lock()
+	defer e.headLock.Unlock()
+	refL1, refL2, _, err := e.SyncRef.RefByL2Num(ctx, nil, &e.Genesis)
+	if err != nil {
+		return err
+	}
+	e.l1Head = refL1
+	e.l2Head = refL2
+	return nil
+}
+
+func (e *EngineDriver) Drive(ctx context.Context, dl Downloader, l1Heads <-chan eth.HeadSignal) ethereum.Subscription {
+	e.driveLock.Lock()
+	defer e.driveLock.Unlock()
+	if e.driveSub != nil {
+		return e.driveSub
+	}
+	e.driveSub = event.NewSubscription(func(quit <-chan struct{}) error {
+		// keep making many sync steps if we can make sync progress
+		hot := time.Millisecond * 30
+		// check on sync regularly, but prioritize sync triggers with head updates etc.
+		cold := time.Second * 8
+		// at least try every minute to sync, even if things are going well
+		max := time.Minute
+
+		syncTicker := time.NewTicker(cold)
+		defer syncTicker.Stop()
+
+		// backoff sync attempts if we are not making progress
+		backoff := cold
+		syncQuickly := func() {
+			syncTicker.Reset(hot)
+			backoff = cold
+		}
+		// exponential backoff, add 10% each step, up to max.
+		syncBackoff := func() {
+			backoff += backoff / 10
+			if backoff > max {
+				backoff = max
+			}
+			syncTicker.Reset(backoff)
+		}
+
+		l2HeadPollTicker := time.NewTicker(time.Second * 14)
+		defer l2HeadPollTicker.Stop()
+
+		onL2Update := func() {
+			// When we updated L2, we want to continue sync quickly
+			syncQuickly()
+			// And we want to slow down requesting the L2 engine for its head (we just changed it ourselves)
+			// Request head if we don't successfully change it in the next 14 seconds.
+			l2HeadPollTicker.Reset(time.Second * 14)
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-quit:
+				return nil
+			case <-l2HeadPollTicker.C:
+				ctx, cancel := context.WithTimeout(ctx, time.Second*4)
+				if err := e.RequestHeadUpdate(ctx); err != nil {
+					e.Log.Error("failed to fetch L2 head info", "err", err)
+				}
+				cancel()
+				continue
```

Comment:
Don't we want to verify that the L2 head of the execution engine is either:
- something we're already aware of
- congruent with the referenced L1 block (i.e. valid) (in the case this is a head that was synced by the execution engine that we didn't set ourselves)

Reply:
We only need to track what the engine says it has to determine if we can go through the happy sync path. We trust our engine to have L2 blocks with correct L1 information (since we inserted it and set it with forkchoice updates). FindSyncStart can handle any out-of-sync cases.

Reply:
But if we're making the assumption that we are the one inserting the L2 blocks, what's the purpose of getting the execution engine to tell us about new heads (which we already know about & track).

I guess the point of this is being a stub for later, when we do have L2-level sync, at which point we will need to validate the block, right?